Internationalization

OkamiKiraa opened the pull request:
- Added SupportedLanguage enum for language management,
- Created English and Polish resource bundles,
- Added language selection functionality on startup,
- Added multiline properties support for menu and help text display,
- Implemented LanguageManager into Main.java file,
- Replaced strings with ResourceBundle messages,
- Added proper error handling for missing property,
- Prepared infrastructure for easy language expansion,

Files added:
- SupportedLanguage.java
- LanguageManager.java  
- messages_en.properties (English)
- messages_pl.properties (Polish)

OkamiKiraa committed:
- Added SupportedLanguage enum for language management,
- Created English and Polish resource bundles,
- Added language selection functionality on startup,
- Added multiline properties support for menu and help text display,
- Implemented LanguageManager into Main.java file,
- Replaced strings with ResourceBundle messages,
- Added proper error handling for missing property,
- Prepared infrastructure

Files added:
- SupportedLanguage.java
- LanguageManager.java
- messages_en.properties (English)
- messages_pl.properties (Polish)
 for easy language expansion,

(cherry picked from commit 6b215757beb56427a70dd22357b3ced8c053c701)

OkamiKiraa committed:
- Added SupportedLanguage enum for language management,
- Created English and Polish resource bundles,
- Added language selection functionality on startup,
- Added multiline properties support for menu and help text display,
- Implemented LanguageManager into Main.java file,
- Replaced strings with ResourceBundle messages,
- Added proper error handling for missing property,
- Prepared infrastructure for easy language expansion,

Files added:
- SupportedLanguage.java
- LanguageManager.java
- messages_en.properties (English)
- messages_pl.properties (Polish)

blackarasu commented on src/Main.java:
```diff
@@ -1,11 +1,16 @@
 import managers.FileManager;
 import managers.StorageHandler;
+import managers.LanguageManager;
+import managers.SupportedLanguage;
```
jak teraz tak widzę, to w sumie fajnie by było ujednolicić formatowanie na google'owe - https://github.com/HPI-Information-Systems/Metanome/wiki/Installing-the-google-styleguide-settings-in-intellij-and-eclipse instrukcja i plik jaki nalezy pobrac jest pod tym linkiem. po tym kliknąć prawym na folder src -> reformat code

blackarasu commented on src/Main.java:
```diff
@@ -37,90 +48,102 @@ public static void main(String[] args) {
+
+        String choice = scanner.nextLine();
+
+        if ("2".equals(choice)) {
```
zamiast takiego ifa lepiej zastosować switcha z default option

OkamiKiraa replied:
Uważam, że na ten moment if ma akceptowalne zastosowanie, gdyż nie ma poszerzonego użycia, jeśli w przyszłości dodane będzie wiele języków / przypadków, wtedy można pomysleć o zaaplikowaniu metody switch albo map

blackarasu replied:
Ma to sens, zamykam

blackarasu commented on src/Main.java:
```diff
@@ -14,19 +19,25 @@ public class Main {
         while (true) {
-            System.out.println(getMenu());
-            System.out.print("Wybierz opcję: ");
+            System.out.println(languageManager.getMessage("menu.display"));
```
zamiast takiego stringa "menu.display" lepiej by stworzyć enumy pod każdy typ wiadomości np. enum Menu z wartościami DISPLAY oraz CHOOSE

nie będę pisał tego przy każdego typu wystąpieniu jako, że jest ich bardzo dużo

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,44 @@
+    private ResourceBundle bundle;
+
+    LanguageManager() {
+
```
niepotrzebna pusta linia

blackarasu commented on src/managers/FileManager.java:
```diff
@@ -15,7 +15,7 @@ public void saveBooks(Library library) {
             }
         } catch (IOException e) {
-            System.out.println("Błąd podczas zapisywania książek: " + e.getMessage());
+            System.out.println("Error while saving books: " + e.getMessage());
```
brakuje wiadomości z LanguageManagera

blackarasu commented on src/managers/FileManager.java:
```diff
@@ -34,7 +34,7 @@ public void loadBooks(Library library) {
             }
         } catch (IOException e) {
-            System.out.println("Błąd podczas ładowania książek: " + e.getMessage());
+            System.out.println("Error while loading books: " + e.getMessage());
```
brakuje wiadomości z LanguageManagera

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,44 @@
+    private static LanguageManager instance;
+    private ResourceBundle bundle;
+
+    LanguageManager() {
```
dodać akcesor private

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,44 @@
+        setLanguage(SupportedLanguage.ENGLISH);
+    }
+
+    public static synchronized LanguageManager getInstance() {
```
póki co nasza appka jest uruchamiana na jednym wątku więc nie potrzeba słówka synchronized

blackarasu commented on src/controller/LibraryController.java:
```diff
@@ -1,17 +1,22 @@
 
     public LibraryController(Library library) {
         this.library = library;
+        this.languageManager = LanguageManager.getInstance();
```
jako, że jest to singleton i zawsze przyjmie taką samą wartość możesz zainicjalizować languageManagera inline w linii 15

blackarasu commented on src/Main.java:
```diff
@@ -14,19 +19,25 @@ public class Main {
+    private static LanguageManager languageManager;
 
     public static void main(String[] args) {
+        languageManager = LanguageManager.getInstance();
```
jako, że jest to singleton i zawsze przyjmie taką samą wartość możesz zainicjalizować languageManagera inline w linii 24

blackarasu commented on src/view/LibraryView.java:
```diff
@@ -8,7 +10,7 @@ public class LibraryView {
     public static String showBooks(Collection<Book> books) {
         if (books.isEmpty()) {
-            return "Brak książek spełniających kryteria.";
+            return LanguageManager.getInstance().getMessage("search.no_results");
```
Czytelniej jest zrobić statyczne pole `private static final LanguageManager languageManager = LanguageManager.getInstance();`

blackarasu commented on src/view/LibraryView.java:
```diff
@@ -8,7 +10,7 @@ public class LibraryView {
 
```
jako, że jest to statyczna klasa to fajnie uniemożliwić "złemu" deweloperowi utworzenie z tej klasy obiektu/ów 
zaleca się użycia w takich klasach prywatnego pustego konstruktora
w tym przypadku `private LibraryView(){}`

ps. taki sam przypadek dla BookView

blackarasu reviewed (changes requested):
zmiany opisane w 12 komentarzach + dostosować BookView.showBook do resourceBundle'i

OkamiKiraa committed:
- UPDATE: added google formatting for better transparency,
- replaced all string keys ("menu.display", "borrow.enter", etc.)  with dedicated enum types for improved type safety and refactoring support,
- refactor Main.java to use enum-based message keys,
- refactor LibraryController.java to use enum-based message keys,
- refactor LibraryView.java to use enum-based message keys,
- updated LanguageManager with getMessage() methods for each enum type,
- LibraryView and BookView are now private with an empty constructor,
- added two missing messages in FileManager.java,
- added missing messages in messages_en.properties (#ADD),
- added missing messages and translation in messages_pl.properties (#ADD),
- fixed missing private accessory in LanguageManager.java,
- fixed empty line in LanguageManager.java,
- removed word 'synchronized' in LanguageManager.java,

blackarasu commented on src/view/BookView.java:
```diff
@@ -3,6 +3,7 @@
+    private BookView(){}
 
     public static String showBook(Book book) {
         return String.format("\"%s\" by %s", book.getTitle(), book.getAuthor());
```
No show book translation

blackarasu commented on src/view/LibraryView.java:
```diff
@@ -1,14 +1,19 @@
 import java.util.Collection;
 
 public class LibraryView {
+    private LibraryView(){}
```
Please use java fields orders convention according to https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+
```
unnecessary blank lane

blackarasu reviewed (commented):
Additionally to get rid of function overloads but to have same functionality add Interface (in this case MessageKey  with `String getKey()` function can be fine) add to each class with suffix MessageKey `implements MessageKey` and add `@Override` above getKey function
then instead of multiple Overloaded function in LanguageManager you can have 2 methods
```java
public String getMessage(MessageKey key) {
    return getMessage(key.getKey());
}

public String getMessage(MessageKey key, Object... args) {
    return getMessage(key.getKey(), args);
}
```

OkamiKiraa committed:
- REUPDATE: added google formatting for better transparency and is now present,
- Library.Java is now reorganized according to java fields convention,
- refactor: replaced methods overloads with MessageKey interface,
- added @Override to all getKey() implementations,
- added missing translation in BookView.java.

blackarasu commented on src/controller/LibraryController.java:
```diff
@@ -1,17 +1,22 @@
 
     public LibraryController(Library library) {
         this.library = library;
+        languageManager = LanguageManager.getInstance();
```
it is not necessary when you assign this value before every constructor of an object in line 14

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,49 @@
+            String message = bundle.getString(key);
+            return String.format(message, args);
+        } catch (Exception e) {
+            return "Missing: " + key;
```
theoretically a missing translation but we can consider that it might also mean that we couldn't find properties file

OkamiKiraa replied:
Let's assume 'theoretically' like you mentioned

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,49 @@
+        try {
+            return bundle.getString(key);
+        } catch (Exception e) {
+            return "Missing: " + key;
```
theoretically a missing translation but we can consider that it might also mean that we couldn't find properties file

OkamiKiraa replied:
Let's assume 'theoretically' like you mentioned

blackarasu commented on src/Main.java:
```diff
@@ -37,90 +54,103 @@ public static void main(String[] args) {
+
+        String choice = scanner.nextLine();
+
+        if ("2".equals(choice)) {
```
Let's be consistent about not having "magic" strings

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,49 @@
+        setLanguage(SupportedLanguage.ENGLISH);
+    }
+
+    public static synchronized LanguageManager getInstance() {
```
as we run application in one thread we don't need synchronized for now

OkamiKiraa replied:
This one is totally bizzare as I refactored it in previous commit - still my bad I guess. I will be more carefull next time

blackarasu commented on src/managers/MessageKey.java:
```diff
@@ -0,0 +1,5 @@
+package managers;
+
+public interface MessageKey {
+    String getKey();
```
be consistent about having blank line before first element in a class/interface + look at enums which use this interface

blackarasu commented on src/Main.java:
```diff
@@ -1,32 +1,49 @@
+    private static LanguageManager languageManager = LanguageManager.getInstance();
 
     public static void main(String[] args) {
+        languageManager = LanguageManager.getInstance();
```
you don't need to assign LanguageManager.getInstance() when you already do it before in line 30

blackarasu commented on src/managers/LanguageManager.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.ResourceBundle;
+
+public class LanguageManager {
+    private static LanguageManager instance;
```
same as MessageKey

blackarasu reviewed (changes requested):
Some small changes about formatting code, some magic string, reassigning singleton and reapplying one thing which was already code reviewed

OkamiKiraa committed:
- removed languageManager = LanguageManager.getInstance(); value in LibraryController.java & Main.java,
- replaced the magic string "2" with simple constant,
- removed word "synchronized" which magically appeared in LanguageManager.java,
- blank lines in MessageKey.java & LanguageManager.java have disappeared.
